pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/blocks/TransferableBlock.java

```diff
@@ -47,6 +48,49 @@ public class TransferableBlock implements Block {
   private BaseDataBlock _dataBlock;
   private List<Object[]> _container;
 
+  /**
+   *  Split a block into multiple block so that each block size is within maxBlockSize.
+   *  Currently, we only support split for row type dataBlock.
+   *  For columnar and metadata data block, we return the original data block.
+   *  For metadata data block, it has to be initialized through bytes instead of data block.
+   *
+   *  When row size is greater than maxBlockSize, we pack each row as a separate block.
+   */
+  private List<TransferableBlock> splitBlock(int maxBlockSize) {
+    List<TransferableBlock> blockChunks = new ArrayList<>();
+    // TODO: Add support for all data block type.
+    if (getType() != BaseDataBlock.Type.ROW) {
+      blockChunks.add(this);
+      return blockChunks;
+    }
+    // TODO: Store row size in bytes inside data block.
+    DataSchema dataSchema = getDataBlock().getDataSchema();
+    int[] columnOffsets = new int[dataSchema.size()];
+    int rowSizeInBytes = DataBlockUtils.computeColumnOffsets(dataSchema, columnOffsets);
+    List<Object[]> chunk = new ArrayList<>();
+    long chunkSize = 0;
+    for (Object[] unit : getContainer()) {
+      if (rowSizeInBytes >= maxBlockSize) {
```

Comment:
if you use `rowSizeInBytes` to split chunks, you know in advance how many rows to write b/c it is fixed. no need to check every row (unit)